Implement a BitField structure

AndrejMitrovic opened the pull request:
There is one problem: We have to rename the `init` method in ocean's `BitArray` into something else like `initialize`, because having `.init` with parameters causes compiler errors in generic code (specifically vibe.d serialization).

So we need to either fix this upstream, or in our fork. I am not sure if `BitArray` was ever used at all in any Sociomantic code (because it lacks all types of attributes like nothrow / @safe / etc), so it might be safe to fix this upstream.

~Requires https://github.com/bpfkorea/agora/pull/415~
Requires https://github.com/sociomantic-tsunami/ocean/issues/781 to be resolved.

AndrejMitrovic committed:
Implement a BitField structure

It's using BitArray as its backing store,
and additionally has serialization and
hashing support.

AndrejMitrovic commented:
The Phobos implementation uses `size_t[]` as its backing store, making it more complex to code for (making it platform-specific I mean). So I opted to use the Ocean implementation instead. The Phobos version might be more battle-tested though, so I'm reconsidering my choice..

AndrejMitrovic commented:
Unblocked and ready for review.

AndrejMitrovic commented:
Update: I've simplified the PR. Previously it was doing byte-level optimizations, but on second thought these are overkill. We can optimize later and just keep it simple and correct now.

Geod24 commented on source/agora/common/BitField.d:
```diff
@@ -0,0 +1,191 @@
+/*******************************************************************************
+
+    Contains a BitField implementation based on Ocean's BitArray struct.
```
```suggestion
    Contains a BitField implementation based on Ocean's BitArray struct.
    
```

Geod24 commented on source/agora/common/BitField.d:
```diff
@@ -0,0 +1,191 @@
+    Contains a BitField implementation based on Ocean's BitArray struct.
+    The choice of Ocean's over Phobos' implementation of BitArray is
+    because the Phobos implementation requires the backing store length to be
+    a multiple of size_t, which makes it platform-specific and hard to code for.
```
Maybe mention that this structure cannot be platform specific because it is stored in the block. Platform specific behavior is not bad if it's not used in data that is exchanged. I'd guess the platform-specifism of Phobos is to make it more performant.

AndrejMitrovic replied:
> I'd guess the platform-specifism of Phobos is to make it more performant.

They probably used size_t without giving it a second thought.

Geod24 commented on source/agora/common/BitField.d:
```diff
@@ -0,0 +1,191 @@
+import std.math;
+
+/// Ditto
+struct BitField
```
```suggestion
public struct BitField
```

Geod24 commented on source/agora/common/BitField.d:
```diff
@@ -0,0 +1,191 @@
+    public void serialize (scope SerializeDg dg) const @trusted
+    {
+        serializePart(this.num_bits, dg);
+        serializePart(cast(ubyte[])this.storage, dg);
```
This breaks on different endianness

AndrejMitrovic replied:
It's a shame we don't have a big endian system in the CI then.

AndrejMitrovic replied:
Are the existing serialization routines safe w.r.t. endianness?

AndrejMitrovic replied:
For example:

```D
dg((cast(ubyte*)&record)[0 .. long.sizeof]);
```

I think this isn't safe?

Geod24 replied:
> Are the existing serialization routines safe w.r.t. endianness?

Probably not.

> It's a shame we don't have a big endian system in the CI then.

I *think* we could set a MIPS one but it'll take and it has lots of dependency, so let's leave that fun for later.

Geod24 reviewed (commented):
A few comments. Also, since the size is dependent on the number of validators, we probably could optimize serialization at the block level, since we will repeat this information.

AndrejMitrovic committed:
Add serialization support for uint[]

The data is serialized in little-endian format,
as this is the most common endianness expected
to run Agora nodes / wallets / tools.

AndrejMitrovic commented:
Some previous attempt at byte-level optimization: https://github.com/AndrejMitrovic/agora/blob/old-bitfield-struct/source/agora/common/BitField.d

Could roll it in later, but I wanted this PR to be simple.

Updated the PR and included `uint[]` serialization as a separate commit.

AndrejMitrovic commented:
~Hold on, I'm doing swapping on a const array. Of course that doesn't work. Not tested because of the `BigEndian` version block. Heheh.~

Fixed.

AndrejMitrovic commented:
Fixed.
